operators/grpc_operators/common/tensor_proto.cpp

```diff
@@ -0,0 +1,167 @@
+}
+
+void TensorProto::gxf_tensor_to_proto(
+    const nvidia::gxf::Entity& gxf_entity,
+    google::protobuf::Map<std::string, ::holoscan::entity::Tensor>* tensor_map) {
+  auto tensors = gxf_entity.findAll<nvidia::gxf::Tensor, 4>();
+  if (!tensors) { throw std::runtime_error("Tensor not found"); }
+
+  for (auto tensor : tensors.value()) {
+    holoscan::entity::Tensor& tensor_proto = (*tensor_map)[tensor->name()];
+    for (uint32_t i = 0; i < (*tensor)->shape().rank(); i++) {
+      tensor_proto.add_dimensions((*tensor)->shape().dimension(i));
+    }
+    switch ((*tensor)->element_type()) {
+      case nvidia::gxf::PrimitiveType::kUnsigned8:
+        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kUnsigned8);
+        copy_data_to_proto<uint8_t>(tensor.value(), tensor_proto);
+        break;
+      case nvidia::gxf::PrimitiveType::kUnsigned16:
+        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kUnsigned16);
+        copy_data_to_proto<uint16_t>(tensor.value(), tensor_proto);
+        break;
+      case nvidia::gxf::PrimitiveType::kFloat32:
+        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kFloat32);
+        copy_data_to_proto<float>(tensor.value(), tensor_proto);
+        break;
+      default:
+        throw std::runtime_error("Unsupported primitive type");
+    }
+  }
+}
+
+template <typename T>
+void TensorProto::copy_data_to_proto(const nvidia::gxf::Handle<nvidia::gxf::Tensor>& tensor,
+                                     ::holoscan::entity::Tensor& tensor_proto) {
+  if ((*tensor).storage_type() == nvidia::gxf::MemoryStorageType::kDevice) {
+    void* in_data_ptr = (*tensor).pointer();
+    size_t data_size = (*tensor).bytes_size();
+    std::vector<T> in_data(data_size);
+    CUDA_TRY(cudaMemcpy(in_data.data(), in_data_ptr, data_size, cudaMemcpyDeviceToHost));
+    tensor_proto.set_data(in_data.data(), data_size);
+    tensor_proto.set_memory_storage_type(holoscan::entity::Tensor::kDevice);
+  } else {
+    tensor_proto.set_data((*tensor).pointer(), (*tensor).size());
+    tensor_proto.set_memory_storage_type(holoscan::entity::Tensor::kSystem);
+  }
+}
+
+void TensorProto::proto_to_gxf_tensor(
+    nvidia::gxf::Entity& gxf_entity,
+    const google::protobuf::Map<std::string, ::holoscan::entity::Tensor>& tensor_map,
+    nvidia::gxf::Handle<nvidia::gxf::Allocator>& allocator) {
+  for (auto tensor_entry : tensor_map) {
+    const holoscan::entity::Tensor& tensor_proto = tensor_entry.second;
+    auto tensor = gxf_entity.add<nvidia::gxf::Tensor>(tensor_entry.first.c_str());
+    if (!tensor) { throw std::runtime_error("Failed to create tensor"); }
+
+    nvidia::gxf::Shape shape({tensor_proto.dimensions().begin(), tensor_proto.dimensions().end()});
+    switch (tensor_proto.primitive_type()) {
+      case holoscan::entity::Tensor::kUnsigned8:
+        tensor.value()->reshape<uint8_t>(shape, nvidia::gxf::MemoryStorageType::kHost, allocator);
```

Comment:
This is always allocating `kHost`. Shouldn't this handle `kDevice` and `kSystem` as well?